File: experiments/adrianco/experiment-7/bookshop/runs/language=clojure_model=claude-opus-4-8-fast/rep2/test/books/api_test.clj
(ns books.api-test
  (:require [clojure.test :refer [deftest testing is use-fixtures]]
            [books.core :as core]
            [books.db :as db]
            [cheshire.core :as json]
            [ring.mock.request :as mock]))

;; A fresh in-memory database (and handler) per test.
(def ^:dynamic *app* nil)

(defn fresh-app-fixture [f]
  (let [ds (db/datasource ":memory:")]
    (db/init! ds)
    (binding [*app* (core/app ds)]
      (f))))

(use-fixtures :each fresh-app-fixture)

(defn- json-body [m]
  (-> (mock/request :post "/books")
      (mock/content-type "application/json")
      (mock/body (json/generate-string m))))

(defn- parse [response]
  (when (seq (:body response))
    (json/parse-string (:body response) true)))

(defn- post-book [m]
  (*app* (json-body m)))

(deftest health-check
  (testing "GET /health returns 200 ok"
    (let [resp (*app* (mock/request :get "/health"))]
      (is (= 200 (:status resp)))
      (is (= "ok" (:status (parse resp)))))))

(deftest create-and-fetch-book
  (testing "POST /books creates a book and returns 201 with an id"
    (let [resp (post-book {:title "Dune" :author "Herbert" :year 1965 :isbn "111"})
          book (parse resp)]
      (is (= 201 (:status resp)))
      (is (some? (:id book)))
      (is (= "Dune" (:title book)))
      (testing "GET /books/{id} returns the created book"
        (let [get-resp (*app* (mock/request :get (str "/books/" (:id book))))]
          (is (= 200 (:status get-resp)))
          (is (= "Herbert" (:author (parse get-resp)))))))))

(deftest validation-rejects-missing-fields
  (testing "POST /books without title/author returns 400"
    (let [resp (post-book {:year 2000})]
      (is (= 400 (:status resp)))
      (is (= #{"title is required" "author is required"}
             (set (:errors (parse resp))))))))

(deftest list-and-filter-by-author
  (testing "GET /books lists all and supports ?author= filter"
    (post-book {:title "A" :author "Asimov"})
    (post-book {:title "B" :author "Herbert"})
    (post-book {:title "C" :author "Asimov"})
    (let [all (parse (*app* (mock/request :get "/books")))]
      (is (= 3 (count all))))
    (let [asimov (parse (*app* (mock/request :get "/books?author=Asimov")))]
      (is (= 2 (count asimov)))
      (is (every? #(= "Asimov" (:author %)) asimov)))))

(deftest update-book
  (testing "PUT /books/{id} updates fields"
    (let [book (parse (post-book {:title "Old" :author "Author"}))
          resp (*app* (-> (mock/request :put (str "/books/" (:id book)))
                          (mock/content-type "application/json")
                          (mock/body (json/generate-string
                                       {:title "New" :author "Author" :year 2020}))))]
      (is (= 200 (:status resp)))
      (is (= "New" (:title (parse resp))))
      (is (= 2020 (:year (parse resp)))))))

(deftest delete-book
  (testing "DELETE /books/{id} removes the book"
    (let [book (parse (post-book {:title "Temp" :author "Author"}))
          del  (*app* (mock/request :delete (str "/books/" (:id book))))]
      (is (= 204 (:status del)))
      (let [get-resp (*app* (mock/request :get (str "/books/" (:id book))))]
        (is (= 404 (:status get-resp)))))))

(deftest unknown-book-returns-404
  (testing "GET /books/{id} for a missing id returns 404"
    (let [resp (*app* (mock/request :get "/books/9999"))]
      (is (= 404 (:status resp))))))

File: experiments/adrianco/experiment-47-gptoss20b/bookshop/runs/agent=hermes-local_language=typescript_model=mlxlocal/mlx-community--gpt-oss-20b-MXFP4-Q8_prompt=neutral_stack=gptoss/rep5-failed/src/__tests__/api.test.ts
import request from 'supertest';
import app from '../index';
import { initDb, db } from '../database';

beforeAll(async () => {
  await initDb();
});

beforeEach(async () => {
  // clear table
  await db.run('DELETE FROM books');
});

afterAll(() => {
  db.close();
});

test('POST /books creates a book', async () => {
  const res = await request(app)
    .post('/books')
    .send({ title: 'Test Book', author: 'Author', year: 2020, isbn: '123' });
  expect(res.status).toBe(201);
  expect(res.body).toMatchObject({ title: 'Test Book', author: 'Author', year: 2020, isbn: '123' });
  expect(res.body.id).toBeDefined();
});

test('GET /books returns list', async () => {
  await request(app).post('/books').send({ title: 'A', author: 'B' });
  const res = await request(app).get('/books');
  expect(res.status).toBe(200);
  expect(res.body.length).toBe(1);
});

test('GET /books?author= filters', async () => {
  await request(app).post('/books').send({ title: 'X', author: 'Same' });
  await request(app).post('/books').send({ title: 'Y', author: 'Other' });
  const res = await request(app).get('/books').query({ author: 'Same' });
  expect(res.body.length).toBe(1);
  expect(res.body[0].author).toBe('Same');
});

test('PUT /books/:id updates', async () => {
  const create = await request(app).post('/books').send({ title: 'Old', author: 'A' });
  const id = create.body.id;
  const upd = await request(app).put(`/books/${id}`).send({ title: 'New' });
  expect(upd.body.title).toBe('New');
  expect(upd.body.author).toBe('A');
});

test('DELETE /books/:id deletes', async () => {
  const create = await request(app).post('/books').send({ title: 'ToDelete', author: 'A' });
  const id = create.body.id;
  const del = await request(app).delete(`/books/${id}`);
  expect(del.status).toBe(204);
  const get = await request(app).get(`/books/${id}`);
  expect(get.status).toBe(404);
});

test('health check', async () => {
  const res = await request(app).get('/health');
  expect(res.status).toBe(200);
  expect(res.body).toEqual({ status: 'ok' });
});

